testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/DefaultGradleDistribution.groovy

```diff
@@ -20,10 +20,52 @@ import org.gradle.api.internal.artifacts.ivyservice.CacheLayout
 import org.gradle.cache.internal.CacheVersion
 import org.gradle.test.fixtures.file.TestDirectoryProvider
 import org.gradle.test.fixtures.file.TestFile
+import org.gradle.tooling.internal.consumer.DefaultGradleConnector
 import org.gradle.util.GradleVersion
 
 class DefaultGradleDistribution implements GradleDistribution {
-    private static final String DISABLE_HIGHEST_JAVA_VERSION = "org.gradle.java.version.disableHighest";
+
+    /**
+     * The java version mapped to the first Gradle version that supports it.
+     *
+     * @see <a href="https://docs.gradle.org/current/userguide/compatibility.html#java_runtime">link</a>
+     */
+    private static final Map<Integer, String> MAX_SUPPORTED_JAVA_VERSIONS = [
+        9: "4.3",
+        10: "4.7",
+        11: "5.0",
+        12: "5.4", // 5.4 officially added support for JDK 12, but it worked before then.
+        13: "6.0",
+        14: "6.3",
+        15: "6.7",
+        16: "7.0",
+        17: "7.3",
+        18: "7.5",
+        19: "7.6",
+        20: "8.3",
+        21: "8.5",
+        22: "8.8",
+        23: "8.10",
+        24: "8.14"
```

Comment:
💅 please add trailing commas to the maps here.

Reply:
We should add Checkstyle or IntelliJ rules to enforce this